### PR TITLE
[ENH] Compute fiber density and fiber spread from ODF using Bingham distributions

### PR DESCRIPTION
**Pull request by CHrlS98**
As discussed with @villalonreina @RafaelNH @mdesco,
This is an attempt at migrating code from [`scilpy`](https://github.com/scilus/scilpy) for fitting Bingham distributions to ODF and computing fiber metrics such as the fiber density and fiber spread [1]. The idea was that this could serve as a basis for [GSoC 2023 Project 2: *Generalized along tract analysis of fiber orientation dispersion*](https://github.com/dipy/dipy/wiki/Google-Summer-of-Code-2023#project-2-generalized-along-tract-analysis-of-fiber-orientation-dispersion).

The PR is WIP. It still misses unit tests and stuff. I would mostly like some feedback at this point:
1. What would be the preferred data structure for representing Bingham distributions and their derived metrics. Right now I simply use python lists of tuples, but maybe we would like to use a numpy array like we do for tensors... It seems clearer as tuples because each parameter of the distribution is a separate output.
2. Do we want to support arrays of SF instead of a single SF at this point? For instance, given some ODF image, we could output a "Bingham image" like it's done in our `scilpy` script [here](https://github.com/scilus/scilpy/blob/master/scripts/scil_fit_bingham_to_fodf.py). Of course it would be cool, but maybe not needed in a first PR.
3. This does not feel like a `reconst` module. It's more of a voxel based analysis or something. Should this go in another module? Is it worth creating a new one?

Checklist
* [x] Write unit tests.
* [x] Support array of SF instead of single SF?

References
[1] Riffert TW, Schreiber J, Anwander A, Knösche TR. Beyond fractional anisotropy: Extraction of bundle-specific structural metrics from crossing fiber models. NeuroImage. 2014 Oct 15;100:176-91.

### CODE REVIEW

**Comment by villalonreina** [comment]
Hi @CHrlS98, thanks for initiating this pull request. Regarding your questions:
1. We can keep it as lists and tuples for now, but at the end we must convert it to a numpy array to build the Nifti file for the full brain image.
2. We will add the feature to loop over the whole brain image, that is to do it over an array of ODFs (á la scilpy).
3. @RafaelNH and I discussed it and we think it should go into the `directions` module.

File: dipy/reconst/bingham.py
Comment (villalonreina) [inline dipy/reconst/bingham.py:20]: Can you change this function's name.? It's a bit confusing because SF is too close to FS which is the name of one of the output quantities. We suggest changing it to `bingham_odf`, because this function computes an ODF based on a Bingham function.

File: dipy/reconst/bingham.py
Comment (villalonreina) [inline dipy/reconst/bingham.py:46]: The same for this name, can you remove the 'sf' from this function name? this is for clarity so that it's not to be confused with the FS output parameter.

File: dipy/reconst/bingham.py
Comment (villalonreina) [inline dipy/reconst/bingham.py:172]: Can you simplify this name to `bingham_fiber_density`?

File: dipy/reconst/bingham.py
Comment (villalonreina) [inline dipy/reconst/bingham.py:220]: The same for this one, can you simplify it to `bingham_fiber_spread`?

**Review by villalonreina — CHANGES_REQUESTED** [review]

**Comment by villalonreina** [comment]
Hi @skoudoro. @CHrlS98 and @RafaelNH and I have been working on this PR. I am trying to do a pull request to Charles' PR but Charles' Dipy fork does not show up in Dipy's Pull request interface. See screenshot attached. Do you know what might be the issue causing this?
Screen Shot 2024-02-26 at 9 09 33 AM

**Comment by skoudoro** [comment]
Hi @villalonreina,

No idea, but when it does not show up, just go directly to his repo and it will show up: https://github.com/CHrlS98/dipy/pulls

**Comment by skoudoro** [comment]
Also, @CHrlS98 @villalonreina,

Release in 2 weeks, I was hoping this will be in, but I am not sure it is ready. 
Please, Can you provide some feedback?

**Comment by CHrlS98** [comment]
@skoudoro
We are actively working on it. @villalonreina made some changes we will integrate to the pull request this week and then @RafaelNH will make a pass on it.

I think Rafael might have a better idea whether 2 weeks is a realistic timeline.

**Comment by skoudoro** [comment]
Thank you for the feedback !

**Comment by villalonreina** [comment]
Thanks @skoudoro. When I go to [https://github.com/CHrlS98/dipy/pulls](url) and I try to submit a PR, Github takes me out of the fork of @CHrlS98 and takes me to the Dipy's main site to make a PR directly to dipy/dipy, not CHrlS98/dipy. Anyway, I think that Charles found a way around. We will try to have this ready asap, hopefully for the next release.

**Comment by RafaelNH** [comment]
I removed the WIP tag since we have a first version of everything

File: dipy/direction/bingham.py
Comment (RafaelNH) [inline dipy/direction/bingham.py:231]: Perhaps we should consider removing this warning here given that this function is being used at a low level, e.g. this if statement is being repeatedly evaluated for all voxels in functions such as "bingham_from_odf" and "bingham_from_sh".

**Review by RafaelNH — CHANGES_REQUESTED** [review]
Just two last comments at my side

File: doc/examples/_valid_examples.toml
Comment (RafaelNH) [inline doc/examples/_valid_examples.toml:114]: I forgot to add the new tutorial here - we have to do that.
Also I didn't delete the old example "bingham_fit_fodf.py". What should we do with it? Integrated it with the new tutorial? Delete it? This is something that we need to discuss.

File: dipy/direction/__init__.py
Comment (skoudoro) [inline dipy/direction/__init__.py:7]: `import *` to avoid. Please import only what you think necessary. 

I will create a PR to remove the one above (`.peaks import *`). I just saw it and this is the only one in the codebase

File: dipy/direction/bingham.py
Comment (skoudoro) [inline dipy/direction/bingham.py:28]: `r` is not necessary, to remove

File: dipy/direction/bingham.py
Comment (skoudoro) [inline dipy/direction/bingham.py:122]: We are slowly moving to keyword only so can you replace this by:  

```
def bingham_fit_odf(odf, sphere, *, npeaks=5, max_search_angle=6,
                    min_sep_angle=60, rel_th=0.1):
```

File: dipy/direction/bingham.py
Comment (skoudoro) [inline dipy/direction/bingham.py:123]: `r` not necessary

File: dipy/direction/bingham.py
Comment (skoudoro) [inline dipy/direction/bingham.py:132]: `, optional`

File: dipy/direction/bingham.py
Comment (skoudoro) [inline dipy/direction/bingham.py:242]: same comment as above some keyword-only

File: dipy/direction/bingham.py
Comment (skoudoro) [inline dipy/direction/bingham.py:257]: `, optional`

File: dipy/direction/bingham.py
Comment (skoudoro) [inline dipy/direction/bingham.py:294]: same comment as above for keyword-only

File: dipy/direction/bingham.py
Comment (skoudoro) [inline dipy/direction/bingham.py:60]: to be consistent with all the codebase, I would recommend using numpy here (`np.cos`).  You use it later your code, so any reason to use the math module here ?

File: doc/examples/bingham_fit_fodf.py
Comment (skoudoro) [inline doc/examples/bingham_fit_fodf.py:24]: old formatting style, Can you update this tutorial by following the new formatting style. Check your other example

File: doc/examples/reconst_bingham.py
Comment (skoudoro) [inline doc/examples/reconst_bingham.py:161]: check in dipy.viz.plotting is there not a function that satisfy your needs. otherwise, create one or update one

File: doc/examples/reconst_bingham.py
Comment (skoudoro) [inline doc/examples/reconst_bingham.py:203]: same comment as above

File: doc/examples/reconst_bingham.py
Comment (skoudoro) [inline doc/examples/reconst_bingham.py:243]: Same comment as above

File: doc/examples/reconst_bingham.py
Comment (skoudoro) [inline doc/examples/reconst_bingham.py:283]: same comment as above

**Review by skoudoro — COMMENTED** [review]
Hi @CHrlS98,

I need to take time to go deeper in this code but thank you for this. 
See below, a very fast review for stuff that catch my eyes.

Also, there are many typo, please, look at codespell CI result to fix them.

**Comment by CHrlS98** [comment]
Thanks for the comments @skoudoro . We had a meeting with @villalonreina and @RafaelNH earlier today to discuss things we want to investigate or modify before merging. We'll get back to you for a second round of reviews once they are done.

File: dipy/direction/bingham.py
Comment (CHrlS98) [reply dipy/direction/bingham.py:61]: Legacy code from scilpy. No there is no reason for not using numpy, I changed it.

File: dipy/direction/bingham.py
Comment (CHrlS98) [reply dipy/direction/bingham.py:269]: I did like we discussed, that is, I added a `_bingham_to_odf` internal method in which the check is not done.

**Comment by CHrlS98** [comment]
@villalonreina @RafaelNH 
So I worked on the code and here are a few comments we might want to discuss in a future meeting.

* __On the fiber density__
I changed the way the fiber density is estimated. Instead of creating a `(theta, phi)` grid with constant theta and phi, I use a DIPY repulsion sphere: this way, we are sure all points are at equal distance and cover the same surface area. The sphere can be passed as a parameter. I also added some tests for the new method.

* __On the maximum separation angle__
I tried lowering it using the tests like we discussed. It would seem there is a difference between `odf_to_bingham` and `_bingham_fit_peak` because for the former I can decrease the angle to 35 while for the later it won't work below 45.

* __`odf_to_bingham` and `bingham_from_odf`__
Both these methods exist, and from my understanding, `bingham_from_odf` takes a whole ODF volume as input whereas `odf_to_bingham` takes a single ODF as input. We might want to change the name because it is quite confusing.

File: doc/examples/reconst_bingham.py
Comment (CHrlS98) [reply doc/examples/reconst_bingham.py]: Created a new `image_mosaic` method in `viz.plotting`.

**Comment by CHrlS98** [comment]
TODO
* [ ] T-design for default `bingham_fiber_density()`
* [ ] `peaks_direction_nl` or subdivide the sphere even more for `_bingham_fit_peak` and `odf_to_bingham` to give the same result in tests.
* [ ] always use multi-voxel methods for single and multi-voxel. Names `bingham_to_sf`, `sh_to_bingham`, `sf_to_bingham`.

**Comment by CHrlS98** [comment]
Hi @RafaelNH, is it possible t-design is slightly biased? I create my sphere this way (it doesn't matter if I subdivide it or not, I tried both) :
```
# default sphere is a subdivision of 45 points spherical t-design (11010 points)
verts = np.loadtxt(get_fnames('t-design'))
sphere = Sphere(xyz=verts).subdivide(4)
```

And then I added a test where I check if the integral of a Bingham distribution along the y-axis is equal to the integral along the x-axis:
```
# Rotating the Bingham distribution should not bias the FD estimate
xfits = [(f0_lobe1, k1, k2, axis0, axis1, axis2),
         (f0_lobe1, k1, k2, axis1, axis2, axis0)]
xpars = _convert_bingham_pars(xfits, 2)
xfd = bingham_fiber_density(xpars)

assert_almost_equal(xfd[0], xfd[1])
```

And sure it's close, but not even that close:
```
E           AssertionError: 
E           Arrays are not almost equal to 7 decimals
E            ACTUAL: 6.6747531900990253
E            DESIRED: 6.690992893100228
```
However when I use a unit icosahedron and subdivide it 5 times (10242 vertices), the test passes. I'll keep the unit icosahedron for now because it is better distributed than repulsion spheres, and add an option to control the level of subdivision.

Is the t-design a whole sphere or a hemisphere? Could it be the reason for this bias?

File: dipy/direction/tests/test_bingham.py
Comment (CHrlS98) [inline dipy/direction/tests/test_bingham.py:32]: @RafaelNH If I replace `peak_dir` by `-peak_dir` I can lower the max angle to lesser than 45 and the test comparing `Mus` and Mus_ref` will still pass.

**Comment by villalonreina** [comment]
@CHrlS98 @RafaelNH I made changes to the documentation, fixed typos and fixed references too. I made a PR to Charles' branch. Also, we should solve the angle (`max_search_angle`) question. I tested 6 degrees and 45 degrees in the example and the 45 degree option looks unrealistic. They both look very different when fitted on a brain image.

**Comment by CHrlS98** [comment]
@villalonreina
If we want to move forward with the PR, should we just make `max_search_angle` a mandatory argument and let the user play with it until we figure it out?

**Comment by villalonreina** [comment]
Hi @CHrlS98 , yes I do think so as well, it should be a mandatory input. I was trying to run validation metrics in July and we still have to figure out what the best option would be. What do you think @RafaelNH?

**Comment by skoudoro** [comment]
Hello @CHrlS98, @villalonreina and @RafaelNH,

By curiosity, what is the status of this work?

**Comment by villalonreina** [comment]
> Hello @CHrlS98, @villalonreina and @RafaelNH,
> 
> By curiosity, what is the status of this work?

Hi @skoudoro. Sorry for the delay. I think that the PR is pretty much ready. We have problems though with one of the inputs to the bingham function, i.e. the search angle for the FOD lobe. We decided to include it as a compulsory argument for the end user, but we don't have a clear recommnedation yet for what it should be. We will all meet to discuss about this issue. I think that we need testing and validation. Not sure if the PR can be merged to master before doing that? What do you think? @RafaelNH @CHrlS98

**Comment by CHrlS98** [comment]
@villalonreina @skoudoro I think all the new functions are tested and documented, so if we just change `max_search_angle` to make it mandatory I think it could be merged. I don't know if I'll have time for this before October however.

@RafaelNH Could you go for a second round of review when you have time?

Thanks!

**Comment by CHrlS98** [comment]
@villalonreina @RafaelNH @skoudoro
`max_search_angle` is now a mandatory argument. I would be comfortable merging the PR even without extensive validation of the `max_search_angle` argument. We have time for recommendations later, and at least the code will already be available to our users.

Review when you have some time, thank you!

**Comment by villalonreina** [comment]
Hi @CHrlS98. Thanks so much for the update. Yes, I agree with this step. `max_search_angle` should be mandatory even though we still don't have a clear recommendation of what it should be. Anything from 6 to 45 grades. @RafaelNH and I will meet in a few weeks and will discuss about the issue to try to resolve it, hopefully with some validation (simulations or phantoms?). We'll keep you posted. @skoudoro is up to you and the rest of the team to decide whether it should be merged like this or not. Thanks.

**Review by skoudoro — COMMENTED** [review]
Hi @CHrlS98,

Thank you for this clean code. Overall, it looks good, you just have to update the citing system since we changed and use a more centralized way for citation. I am ok to merge as it is after your update.

I agree that the recommendation can be done  later.

Thanks to @villalonreina and @RafaelNH  for the reviews!

File: dipy/direction/tests/test_bingham.py
Comment (CHrlS98) [reply dipy/direction/tests/test_bingham.py]: I instead added an argument `legacy` to `sh_to_bingham` and set it to false in tests to avoid the warning.

**Comment by CHrlS98** [comment]
@skoudoro

Thank you for the feedback. I answered your review comments.

Let me know if anything else needs changing.

File: doc/examples/reconst_bingham.py
Comment (skoudoro) [inline doc/examples/reconst_bingham.py:45]: I mean, we follow the rst rules: https://docutils.sourceforge.io/docs/user/rst/quickstart.html#sections

so, it should be this (underline same size as the title):
```rst
###############################################################################
# Step 1. ODF estimation
# ======================
```

Can you update everywhere

File: doc/examples/reconst_bingham.py
Comment (skoudoro) [inline doc/examples/reconst_bingham.py:95]: same comment as above

File: doc/examples/reconst_bingham.py
Comment (skoudoro) [inline doc/examples/reconst_bingham.py:174]: same comment as above

**Review by skoudoro — COMMENTED** [review]
Thank you for the update @CHrlS98,

See below the answer to your question

File: doc/examples/reconst_bingham.py
Comment (CHrlS98) [reply doc/examples/reconst_bingham.py]: Alright done!

File: dipy/direction/bingham.py
Comment (villalonreina) [reply dipy/direction/bingham.py]: @CHrlS98 @skoudoro I think this is the input to make mandatory instead of optional, is that right?

Comment (CHrlS98) [reply dipy/direction/bingham.py]: Yes, I removed the default value. It must be specified by the user.

**Comment by skoudoro** [comment]
Hi @CHrlS98 ,

Thank you for the  lastupdate. 

Code format CI is still failing as you can see here: https://github.com/dipy/dipy/actions/runs/11389486311/job/31688883284?pr=2826#step:4:137

A good way to check that locally is to use pre-commit:
```
pip install pre-commit 
pre-commit install
pre-commit run --files your_files.py
```

Also, build_docs CI's is failing also. here are the warnings, but you can test it locally by doing in the `docs`
 folder: `make - C . html-no-examples` (you might have to update the requirements, check doc-requirements.txt)

```python
/opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/numpydoc/docscrape.py:421: UserWarning: Unknown section Return in the docstring of sf_to_bingham in /opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/dipy/direction/bingham.py.
  self[section] = content
/opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/numpydoc/docscrape.py:421: UserWarning: Unknown section Return in the docstring of sh_to_bingham in /opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/dipy/direction/bingham.py.
  self[section] = content
/opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/numpydoc/docscrape.py:384: UserWarning: potentially wrong underline length... 
Note: 
---- in 
Overall concentration parameters for an ODF peak.
... in the docstring of kappa_total_lobe in /opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/dipy/direction/bingham.py.
  if not self._is_at_section():
```

**Comment by CHrlS98** [comment]
@skoudoro 
Remaining doc error seems to be
```
/opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/dipy/direction/bingham.py:docstring of dipy.direction.bingham.BinghamMetrics.kappa_total_lobe:4: ERROR: Too many autonumbered footnote references: only 0 corresponding footnotes available. [docutils]
/opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/dipy/direction/bingham.py:docstring of dipy.direction.bingham.BinghamMetrics.kappa_total_lobe:4: ERROR: Unknown target name: "footcite-tariq2016". [docutils]
/opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/dipy/direction/bingham.py:docstring of dipy.direction.bingham.BinghamMetrics.odi_total_lobe:2: ERROR: Unknown target name: "footcite-tariq2016". [docutils]
```

I can't figure out why this particular reference won't work. Would you happen to have a solution?

File: dipy/direction/bingham.py
Comment (skoudoro) [inline dipy/direction/bingham.py:574]: I believe the reason is `footcite` in a short description.  it can not be in a short description. so splitting short and long description like below should work. (I did not try, I can test this later tonight)

```python
"""Return Orientation Dispersion Index (ODI) for an ODF lobe.

The ODF lobe is from the overall concentration parameter (k_total).Defined by equation 20 in :footcite:p:`Tariq2016`.

References
----------
.. footbibliography::
"""
```

File: dipy/direction/bingham.py
Comment (skoudoro) [inline dipy/direction/bingham.py:547]: I believe it is good but disturbed by the other one. See my other comment.

File: dipy/direction/bingham.py
Comment (CHrlS98) [reply dipy/direction/bingham.py]: Thanks, it appears to have worked.

**Comment by skoudoro** [comment]
Hi @arokem,

Can you have a quick look at this PR before going ahead for merging?  another pair of eye is needed 😅 
@Atharva-Shah-2298 have checked it also. 

Thank you in advance

**Comment by arokem** [comment]
Will do. Give me a day or two

**Comment by skoudoro** [comment]
Hi @arokem,

Do you have any update on this? Can we go ahead ? Thank you for your feedback

**Comment by arokem** [comment]
Sorry - I got distracted by many other things... 

Right off the bat, an answer to a question that was asked at the very top: I do think this should go into the reconst module. Seems unjustified to start a new module for this.

**Comment by arokem** [comment]
Even the example (looks great!) is called reconst_bingham, suggesting that this is the module to look into.

File: dipy/direction/bingham.py
Comment (arokem) [inline dipy/direction/bingham.py:292]:
```suggestion
        Map marking the coordinates in the data that should be analyzed.
        Default (None) means all voxels in the volume will be analyzed.
```

File: dipy/direction/bingham.py
Comment (arokem) [inline dipy/direction/bingham.py:491]:
```suggestion
        Any metric with values for nl ODF lobes.
```

File: dipy/direction/bingham.py
Comment (arokem) [inline dipy/direction/bingham.py:691]:
```suggestion
    # Bingham parameters stored in an ndarray with shape:
```

**Review by arokem — COMMENTED** [review]
This is great work! I look forward to using it. 

Other than my comment about whether to include this in the `reconst` module (I think so, but not adamant), I had just a few small comments.

**Comment by skoudoro** [comment]
Thanks @arokem!

Good catch! +1 for having this in `reconst` module

**Comment by skoudoro** [comment]
Hi @CHrlS98,

it would be great if you could address ariel comment today or tomorrow morning 😅

**Review by skoudoro — APPROVED** [review]
Thank you for this great work @CHrlS98!

Thank you for all the reviews @RafaelNH, @villalonreina, @arokem

merging

**Pull request merged**
